DAOS-9820 test: make large datamover tests more configurable

daltonbohning opened the pull request:
Test-tag: datamover,-iosysadmin,-dm_posix_meta_entry
Skip-test: true
Skip-build: true

- Add `KeywordParameter` class that accepts a mapping of replaced values
- `data_mover_test_base`
  - Support ppn for various datamover utils
  - Use `KeywordParameter` for `posix_root`
- `large_dir` and `large_file`
  - Specify ppn instead of np
  - Dual-engine and move to HW Medium
  - Specify posix_root in the yaml
- `large_dir`
  - Remove duplicate variants in favor of the larger
  - Use EC object class by default, similar to `large_file`

Signed-off-by: Dalton Bohning <dalton.bohning@intel.com>

daosbuild1 commented on src/tests/ftest/util/data_mover_test_base.py:
```diff
@@ -1105,7 +1105,7 @@
             expected_err (list, optional): substrings expected in stderr
             processes (int, optional): number of mpi processes.
-                defaults to self.dcp_processes
+                Defaults to np for correspending tool.
```
```suggestion
                Defaults to np for corresponding tool.
```

daosbuild1 commented on src/tests/ftest/util/data_mover_utils.py:
```diff
@@ -127,7 +127,11 @@ def run(self, processes, job_manager):
         job_manager.assign_hosts(self.hosts, self.tmp)
-        job_manager.assign_processes(processes)
+        if ppn is None:
+            manager.assign_processes(processes)
```
(pylint-undefined-variable) Undefined variable 'manager'

daosbuild1 commented on src/tests/ftest/util/data_mover_utils.py:
```diff
@@ -127,7 +127,11 @@ def run(self, processes, job_manager):
+        if ppn is None:
+            manager.assign_processes(processes)
+        else:
+            manager.ppn.update(ppn, 'mpirun.ppn')
```
(pylint-undefined-variable) Undefined variable 'manager'

daosbuild1 commented on src/tests/ftest/util/data_mover_utils.py:
```diff
@@ -127,7 +127,11 @@ def run(self, processes, job_manager):
+            manager.assign_processes(processes)
+        else:
+            manager.ppn.update(ppn, 'mpirun.ppn')
+            manager.processes.update(None, 'mpirun.np')
```
(pylint-undefined-variable) Undefined variable 'manager'

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8662/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/1a76ab8dcd7b7d8c570458bb0ed06753987d3a84/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1289/1000)

[src/tests/ftest/util/data_mover_test_base.py:24](https://github.com/daos-stack/daos/blob/1a76ab8dcd7b7d8c570458bb0ed06753987d3a84/src/tests/ftest/util/data_mover_test_base.py#L24):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (21/20)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/1/execution/node/129/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/2db355b338db2ecfffc6e5d8e72516a91da427f7/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1289/1000)

[src/tests/ftest/util/data_mover_test_base.py:24](https://github.com/daos-stack/daos/blob/2db355b338db2ecfffc6e5d8e72516a91da427f7/src/tests/ftest/util/data_mover_test_base.py#L24):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (21/20)

daosbuild1 commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -245,6 +255,30 @@ def update(self, value, name=None, append=False):
 
+class KeywordParameter(BasicParameter):
+    """A class for parameters whose values are read from a yaml file.
+    
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion

```

daosbuild1 commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -245,6 +255,30 @@ def update(self, value, name=None, append=False):
+    """
+
+    KEYWORD_DICT = {
+        "^python://": (lambda s: eval(re.sub(r'^python://(.*)$', r'str(\1)', s)))
```
(pylint-eval-used) Use of eval

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8662/3/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/b9e77a20c130c1613f5594623cb4d35d4d0d2465/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1289/1000)

[src/tests/ftest/util/data_mover_test_base.py:25](https://github.com/daos-stack/daos/blob/b9e77a20c130c1613f5594623cb4d35d4d0d2465/src/tests/ftest/util/data_mover_test_base.py#L25):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (21/20)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/3/execution/node/129/log

daosbuild1 commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -245,6 +256,30 @@ def update(self, value, name=None, append=False):
+    }
+
+    @BasicParameter.value.getter
+    def value(self):
```
(pylint-invalid-overridden-method) Method 'value' was expected to be 'property', found it instead as 'method'

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8662/4/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/9acb2229fd626e69a31d598568c35842a7f388a2/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1289/1000)

[src/tests/ftest/util/data_mover_test_base.py:25](https://github.com/daos-stack/daos/blob/9acb2229fd626e69a31d598568c35842a7f388a2/src/tests/ftest/util/data_mover_test_base.py#L25):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (21/20)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/4/execution/node/129/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/71771b64dca0d4aa15eeb4b1de49b94e245c17ba/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1289/1000)

[src/tests/ftest/util/data_mover_test_base.py:25](https://github.com/daos-stack/daos/blob/71771b64dca0d4aa15eeb4b1de49b94e245c17ba/src/tests/ftest/util/data_mover_test_base.py#L25):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (21/20)

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/176ed2fb9dbd2841128fd67388b5fcb53217e5dd/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1292/1000)

[src/tests/ftest/util/data_mover_test_base.py:25](https://github.com/daos-stack/daos/blob/176ed2fb9dbd2841128fd67388b5fcb53217e5dd/src/tests/ftest/util/data_mover_test_base.py#L25):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (21/20)

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/3099ccd82d03cc0df37c8930b3c6a0b48d9e953a/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1164/1000)

[src/tests/ftest/util/data_mover_test_base.py:26](https://github.com/daos-stack/daos/blob/3099ccd82d03cc0df37c8930b3c6a0b48d9e953a/src/tests/ftest/util/data_mover_test_base.py#L26):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (22/20)

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/a9c514c4749f3242d536522ceef1ac437049c71b/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1164/1000)

[src/tests/ftest/util/data_mover_test_base.py:26](https://github.com/daos-stack/daos/blob/a9c514c4749f3242d536522ceef1ac437049c71b/src/tests/ftest/util/data_mover_test_base.py#L26):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (22/20)

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/9/execution/node/394/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/9/execution/node/326/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/9/execution/node/383/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/9/execution/node/380/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/a9c514c4749f3242d536522ceef1ac437049c71b/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1164/1000)

[src/tests/ftest/util/data_mover_test_base.py:26](https://github.com/daos-stack/daos/blob/a9c514c4749f3242d536522ceef1ac437049c71b/src/tests/ftest/util/data_mover_test_base.py#L26):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (22/20)

daltonbohning commented:
3 known test failures: https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-8662/10/tests
https://daosio.atlassian.net/browse/DAOS-10344

Looking at the log for `large_file`: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8662/10/artifact/Functional%20Hardware%20Medium/datamover/large_file.py/job.log
ppn is updated for IOR:
```
Updated param mpirun.ppn => 20
Updated param mpirun.np => None
...
mpirun ior -ppn 20
```
ppn is updated for `dcp`:
```
Updated param mpirun.ppn => 32
Updated param mpirun.np => None
...
mpirun dcp -ppn 32
```

daltonbohning commented:
@phender This PR introduces the `KeywordParameter` as we discussed, though I wonder if a better name might be `MappedParameter` ?

phender commented on src/tests/ftest/util/data_mover_test_base.py:
```diff
@@ -73,19 +76,31 @@ def __init__(self, *args, **kwargs):
+        self.ddeserialize_np = None
+
+        # Root directory for POSIX paths
+        self.posix_root = KeywordParameter(None, mapping={
+            'self.workdir': self.workdir,
+            'self.tmp': self.tmp
+        })
```
Minor cosmetic suggestion + adding the default:
```suggestion
        posix_root_map = {'self.workdir': self.workdir, 'self.tmp': self.tmp}
        self.posix_root = KeywordParameter(None, mapping=posix_root_map, default=self.tmp)
```

phender commented on src/tests/ftest/util/data_mover_test_base.py:
```diff
@@ -114,23 +129,23 @@ def setUp(self):
+        self.ddeserialize_np = self.params.get("np", "/run/ddeserialize/*", self.processes)
+        self.ddeserialize_ppn = self.params.get("ppn", "/run/ddeserialize/*", self.ppn)
+
+        self.posix_root.update(self.params.get("posix_root", "/run/datamover/*", self.tmp))
```
Assign the default with creating the `KeywordParameter` object and then use it's `get_yaml_value()` method to read the test yaml:
```suggestion
        self.posix_root.get_yaml_value("posix_root", self, "/run/datamover/*")
```

phender commented:
> @phender This PR introduces the `KeywordParameter` as we discussed, though I wonder if a better name might be `MappedParameter` ?

That works too, and its probably more descriptive.

daosbuild1 commented on src/tests/ftest/util/data_mover_test_base.py:
```diff
@@ -20,6 +20,7 @@
 import re
 import ctypes
 from general_utils import create_string_buffer
+from command_utils_base import MappedParameter
```
(pylint-no-name-in-module) No name 'MappedParameter' in module 'command_utils_base'

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8662/11/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/30a497164a5eae76d2ed58003c434e1e4b9ffa39/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1162/1000)

[src/tests/ftest/util/data_mover_test_base.py:26](https://github.com/daos-stack/daos/blob/30a497164a5eae76d2ed58003c434e1e4b9ffa39/src/tests/ftest/util/data_mover_test_base.py#L26):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (22/20)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8662/11/execution/node/114/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/93c606ffb689d453b68ad6b8a143d9894aac60f9/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1162/1000)

[src/tests/ftest/util/data_mover_test_base.py:26](https://github.com/daos-stack/daos/blob/93c606ffb689d453b68ad6b8a143d9894aac60f9/src/tests/ftest/util/data_mover_test_base.py#L26):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (22/20)

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/22002b335ada2d8e29e3077ef59b9082e7326150/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1162/1000)

[src/tests/ftest/util/data_mover_test_base.py:26](https://github.com/daos-stack/daos/blob/22002b335ada2d8e29e3077ef59b9082e7326150/src/tests/ftest/util/data_mover_test_base.py#L26):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (22/20)

daltonbohning commented:
obj_large_posix and serial_large_posix are known failures:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-8662/13/tests/
https://daosio.atlassian.net/browse/DAOS-10344

daltonbohning commented:
Ping @phender

dinghwah reviewed (approved):
LGTM

daltonbohning commented:
> LGTM. Please address the 3 failures which might not related to the commit.

Thanks @dinghwah. The 3 failures are known: https://daosio.atlassian.net/browse/DAOS-10344

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/data_mover_test_base.py:1](https://github.com/daos-stack/daos/blob/a01febee76024ca35c2881ebe8ded07ae33028c2/src/tests/ftest/util/data_mover_test_base.py#L1):
(pylint-too-many-lines) Too many lines in module (1162/1000)

[src/tests/ftest/util/data_mover_test_base.py:26](https://github.com/daos-stack/daos/blob/a01febee76024ca35c2881ebe8ded07ae33028c2/src/tests/ftest/util/data_mover_test_base.py#L26):
(pylint-too-many-instance-attributes) Too many instance attributes (36/30)
(pylint-too-many-public-methods) Too many public methods (22/20)

[src/tests/ftest/util/data_mover_test_base.py:360](https://github.com/daos-stack/daos/blob/a01febee76024ca35c2881ebe8ded07ae33028c2/src/tests/ftest/util/data_mover_test_base.py#L360):
(pylint-arguments-renamed) Parameter 'chunk_size' has been renamed to 'pool' in overridden 'DataMoverTestBase.create_cont' method
(pylint-arguments-renamed) Parameter 'properties' has been renamed to 'use_dfuse_uns' in overridden 'DataMoverTestBase.create_cont' method
